nomad/eval_broker.go

```diff
@@ -876,6 +876,14 @@ func (b *EvalBroker) EmitStats(period time.Duration, stopCh <-chan struct{}) {
 	}
 }
 
+// HasBlockedEval is used to check if the eval broker already has a blocked eval for this job.
+func (b *EvalBroker) HasBlockedEval(namespacedJobID structs.NamespacedID) bool {
+	b.l.Lock()
+	defer b.l.Unlock()
+	_, ok := b.blocked[namespacedJobID]
+	return ok
+}
```

Comment:
Node update evals always come in with their priority set to the job priority. Maybe we should check if the blocked eval has a lower priority here as well, so that we can bump-over an eval that was submitted by a user with a low priority?

Reply:
My initial concerns around this idea are that things like `SnapshotIndex` and `NodeModifyIndex` aren't getting evals with updated values, but `NodeModifyIndex` is just informational (never consumed by Nomad) and the `SnapshotIndex` is set at _scheduling_ time so there's nothing for us to screw up with that either -- none of the evals in the `PendingEvals` list we call "blocked" here will have `SnapshotIndex` set either.

Reply:
Great analysis. I had assumed `{Job,Node}ModifyIndex` would be used for the minimum index of the state store when initially dequeuing the eval in the worker, but that doesn't appear to be the case. The index is the max `ModifyIndex` of _all evals for the job._ :bangbang: 

(That actually seems like an overly pessimistic index to use as it ensures the oldest eval for a job evaluates against the same index as the _latest_ eval for a job... why even bother processing those subsequent evals then since there's no guarantee the state will be later than when the first eval was committed? If a new eval comes in in the mean time, _it_ would need to get processed.)

Unfortunately I think this logic poses a problem for your otherwise fantastic patch: you will no longer be creating an eval with a potentially higher `ModifyIndex` than all of the blocked evals for the job. So the job may end up not getting evaluated against the latest index... which is particularly a problem if the whole point is to reevaluate the job against _this_ index of the node.

To try to illustrate: The old behavior ("Blocked" means "dupe pending eval for a job" -- BlockedEvals don't factor in here AFAICT)

1. Eval E1 for Job J1 created @ Index 10
2. Eval E2 for Job J1 created and blocked @ Index 20
3. Worker dequeues E1, gets a WaitIndex of **20**
4. Node.Update occurs, updates Node and creates Eval E3 for J1 at Index 30
5. Worker dequeues E2, gets a WaitIndex of **30**
6. Worker dequeues E3, gets a WaitIndex of **30**

With this PR:

1. 1. Eval E1 for Job J1 created @ Index 10
2. Eval E2 for Job J1 created and blocked @ Index 20
3. Worker dequeues E1, gets a WaitIndex of **20**
4. Node.Update occurs, updates Node at index 30, **sees E2 blocked and skips eval creation**
5. Worker dequeues E2, gets a WaitIndex of **20**

So we may never evaluate the node at its latest ModifyIndex (30).

I could be wrong though... lots of moving parts here. Honestly I'm not sure why we calculate WaitIndex the way we do... using `Eval.{Job,Node}ModifyIndex` (based on `TriggeredBy`) initially seems *correct.* Using `max(Eval.ModifyIndex, Eval.SnapshotIndex)` always seems safe and potentially optimal though (I would assume Eval.ModifyIndex is always >= its `{Job,Node}ModifyIndex`)

I almost want to remove `{Job,Node}ModifyIndex` if they're not used. I can't imagine them being useful for end users. :shrug: 

If my assessment of the current behavior is correct, then obviously the eval you're skipping creating is useless to actually process, but it *is* necessary to ensure the modify index is high enough.

Reply:
Yeah. ☹️ I hit this problem from a different direction last night. 

I was trying to work up a model of the schedulers/broker/raft inspired by the [aws s3 lightweight formal methods](https://www.amazon.science/publications/using-lightweight-formal-methods-to-validate-a-key-value-storage-node-in-amazon-s3) paper (which is an excellent read, btw). But I needed an invariant to test: we want any change to cluster state (i.e. state at a given index) to be eventually reflected in the last evaluation processed. But without writing the evaluation to raft we don't have a valid index to compare against anywhere in the scheduler.

So for some reason I had it in my head we only had a wait index concept in the planner but it's obvious in retrospect that we'd need it in the scheduler too.

Reply:
> (That actually seems like an overly pessimistic index to use as it ensures the oldest eval for a job evaluates against the same index as the _latest_ eval for a job... why even bother processing those subsequent evals then since there's no guarantee the state will be later than when the first eval was committed? If a new eval comes in in the mean time, it would need to get processed.)

This observation could inform an approach to shed duplicates from the eval broker, even if we can't prevent them from getting written to raft. Don't try to avoid adding them to the `blocked` list of `PendingEvals`. But if we get an ack for an eval at `ModifyIndex = n`, then I think we can immediately discard all evals in `PendingEvals` for that job with `ModifyIndex` < `n` and batch cancel them for efficiency.

Reply:
Hacking on this idea over in https://github.com/hashicorp/nomad/compare/main..f-evaluation-load-shedding-take-2 but will update this PR with the results once it's at all suitable for discussion.

Reply:
Something I've run into here is that when we `heap.Pop` the `blocked` queue, we're always getting the oldest evaluation instead of the newest one. So when it later gets ack'd, it's never newer than the others. The `PendingEvals` type sorts by Priority and then CreateIndex FIFO, which is fine for `pending` (we probably need to do that for fairness across jobs). But for the `blocked` queue we want to sort by Priority but CreateIndex LIFO for this idea to work. Will have to split the types and see where that goes next.

Reply:
> we're always getting the oldest evaluation instead of the newest one. So when it later gets ack'd, it's never newer than the others.

While this is true, I don't think it matters because `WaitIndex` given to workers as the min index to work from is calculated to be the _max modify index for all evals for the job:_ https://github.com/hashicorp/nomad/blob/v1.4.0-beta.1/nomad/eval_endpoint.go#L195-L208

So for a given job it doesn't matter which of its evals we process first, we'll always process it against the latest known index at the time of dequeue. I think when completing an eval we could go ahead and cancel all other evals for the same job with a <= ModifyIndex to the plan's _SnapshotIndex_ as that's guaranteed to be >= the WaitIndex.

Put another way, if you have the following evals for the same job:

1. node-update @ 100
2. preemption @ 200

Then even though the `node-update @ 100` is dequeued, the `WaitIndex` will be `200`. The resulting Plan's `SnapshotIndex` will be >= `200` so if the `node-update @ 100` eval is complete, the `preemption @ 200` eval can be canceled as well.

`Oldest Eval ModifyIndex <= WaitIndex at Dequeue time <= Worker's Snapshot Index`

Reply:
Ok, that behavior doesn't show up at a unit testing level because we're not running a real worker, just making `Dequeue` and `Ack` calls, but that eliminates some extra code I've had to write here. Just need to update the tests to match that behavior and I think I've actually got this working.